packages/shorebird_cli/lib/src/commands/patch/macos_patcher.dart

```diff
@@ -0,0 +1,446 @@
+
+    final File patchFile;
+    if (useLinker && await aotTools.isGeneratePatchDiffBaseSupported()) {
+      final patchBaseProgress = logger.progress('Generating patch diff base');
+      final analyzeSnapshotPath = shorebirdArtifacts.getArtifactPath(
+        artifact: ShorebirdArtifact.analyzeSnapshotMacOS,
+      );
+
+      final File patchBaseFile;
+      try {
+        // If the aot_tools executable supports the dump_blobs command, we
+        // can generate a stable diff base and use that to create a patch.
+        patchBaseFile = await aotTools.generatePatchDiffBase(
+          analyzeSnapshotPath: analyzeSnapshotPath,
+          releaseSnapshot: releaseArtifactFile,
+        );
+        patchBaseProgress.complete();
+      } catch (error) {
+        patchBaseProgress.fail('$error');
+        throw ProcessExit(ExitCode.software.code);
+      }
+
+      patchFile = File(
+        await artifactManager.createDiff(
+          releaseArtifactPath: patchBaseFile.path,
+          patchArtifactPath: patchBuildFile.path,
+        ),
+      );
+    } else {
+      patchFile = patchBuildFile;
+    }
+
+    final patchFileSize = patchFile.statSync().size;
+    final privateKeyFile = argResults.file(CommonArguments.privateKeyArg.name);
+    final hash = sha256.convert(patchBuildFile.readAsBytesSync()).toString();
+    final hashSignature = privateKeyFile != null
+        ? codeSigner.sign(
+            message: hash,
+            privateKeyPemFile: privateKeyFile,
+          )
+        : null;
+
+    // TODO(bryanoltman): support x86_64
+    return {
+      Arch.arm64: PatchArtifactBundle(
+        arch: 'aarch64',
+        path: patchFile.path,
+        hash: hash,
+        size: patchFileSize,
+        hashSignature: hashSignature,
+      ),
+    };
+  }
+
+  @override
+  Future<String> extractReleaseVersionFromArtifact(File artifact) async {
+    final appPath = artifactManager.getMacOSAppDirectory()?.path;
+    if (appPath == null) {
+      logger.err('Unable to find .app directory');
+      throw ProcessExit(ExitCode.software.code);
+    }
+
+    final plistFile = File(p.join(appPath, 'Contents', 'Info.plist'));
+    if (!plistFile.existsSync()) {
+      logger.err('No Info.plist file found at ${plistFile.path}.');
+      throw ProcessExit(ExitCode.software.code);
+    }
+
+    final plist = Plist(file: plistFile);
+    try {
+      return plist.versionNumber;
+    } catch (error) {
+      logger.err(
+        'Failed to determine release version from ${plistFile.path}: $error',
+      );
+      throw ProcessExit(ExitCode.software.code);
+    }
+  }
+
+  @override
+  Future<CreatePatchMetadata> updatedCreatePatchMetadata(
+    CreatePatchMetadata metadata,
+  ) async =>
+      metadata.copyWith(
+        linkPercentage: lastBuildLinkPercentage,
+        environment: metadata.environment.copyWith(
+          xcodeVersion: await xcodeBuild.version(),
+        ),
+      );
+
+  Future<_LinkResult> _runLinker({
```

Comment:
We should probably create a `Linker` class that contains this shared logic instead of duplicating on both iOS and MacOS.

Reply:
Yep, I have a TODO earlier in this file on line 33 for the `_LinkResult` typedef.